kinesis/src/main/scala/akka/stream/alpakka/kinesis/scaladsl/KinesisSchedulerSource.scala

```diff
@@ -0,0 +1,103 @@
+import software.amazon.kinesis.exceptions.ShutdownException
+import software.amazon.kinesis.processor.ShardRecordProcessorFactory
+import software.amazon.kinesis.retrieval.KinesisClientRecord
+
+import scala.collection.immutable
+import scala.concurrent.Future
+
+object KinesisSchedulerSource {
+
+  /**
+   * Source that spawns a single Kinesis KCL Scheduler capable of consuming records from multiple Kinesis
+   * shards in coordination with other KCL Scheduler instances connected to the same stream. Only a single
+   * KCL Scheduler instance can own the lease of a shard at any given time. The Kinesis service assigns each
+   * shard lease among the available instances.
+   *
+   * Every message emitted downstream belongs to a shard. Messages can be committed to save the progress of
+   * the client consuming that shard.
+   */
+  def apply(
+      schedulerBuilder: ShardRecordProcessorFactory => Scheduler,
+      settings: KinesisSchedulerSourceSettings = KinesisSchedulerSourceSettings.defaultInstance
+  ): Source[CommittableRecord, Future[Scheduler]] =
+    Source
+      .setup { (mat, _) =>
+        import mat.executionContext
+        Source
+          .fromGraph(new KinesisSchedulerSourceStage(settings, schedulerBuilder))
+      }
+      .mapMaterializedValue(_.flatMap(identity)(ExecutionContexts.sameThreadExecutionContext))
+
+  def sharded(
+      schedulerBuilder: ShardRecordProcessorFactory => Scheduler,
+      settings: KinesisSchedulerSourceSettings = KinesisSchedulerSourceSettings.defaultInstance
+  ): SubFlow[CommittableRecord, Future[Scheduler], Source[CommittableRecord, Future[Scheduler]]#Repr, RunnableGraph[
+    Future[Scheduler]
+  ]] =
+    apply(schedulerBuilder, settings)
+      .groupBy(MAX_KINESIS_SHARDS, _.processorData.shardId)
+
+  def checkpointRecordsFlow(
+      settings: KinesisSchedulerCheckpointSettings = KinesisSchedulerCheckpointSettings.defaultInstance
+  ): Flow[CommittableRecord, KinesisClientRecord, NotUsed] =
+    Flow[CommittableRecord]
+      .groupBy(MAX_KINESIS_SHARDS, _.processorData.shardId)
+      .groupedWithin(settings.maxBatchSize, settings.maxBatchWait)
+      .via(GraphDSL.create() { implicit b =>
+        import GraphDSL.Implicits._
+        val split =
+          b.add(scaladsl.Broadcast[immutable.Seq[CommittableRecord]](2))
+        val join = b.add(Zip[Done, immutable.Seq[CommittableRecord]])
+        val result = b.add(Flow[KinesisClientRecord])
+
+        split
+          .out(0)
+          .map(_.max)
+          .via(
+            Flow[CommittableRecord]
+              .map {
+                case record if record.canBeCheckpointed =>
+                  record
+                    .tryToCheckpoint()
+                    .recover({
+                      case _: ShutdownException => Done
+                    })
+                    .get
+                case _ => Done
+              }
+              .addAttributes(Attributes(ActorAttributes.IODispatcher))
```

Comment:
If you move this to a `ShardProcessor` object, the semaphore and the IO dispatcher show together and are easier to understand.